src/main/java/org/thingsboard/tools/service/device/BaseDeviceAPITest.java

```diff
@@ -166,41 +195,232 @@ public void subscribeWebSockets() {
         int cmdId = 1;
         for (Map.Entry<String, SubscriptionData> entry : deviceMap.entrySet()) {
             SubscriptionData data = entry.getValue();
-            data.setSubscriptionId(cmdId);
             subscribeToWebSocket(data.getDeviceId(), cmdId);
+            data.setSubscriptionId(cmdId);
             cmdId++;
         }
     }
 
+    @Override
+    public void warmUpDevices(final int publishTelemetryPause) throws InterruptedException {
+        restClient.login(username, password);
+        log.info("Warming up {} devices...", deviceCount);
+        CountDownLatch connectLatch = new CountDownLatch(deviceCount);
+        AtomicInteger totalWarmedUpCount = new AtomicInteger();
+
+        int halfSize = deviceCount / 2;
```

Comment:
I would suggest use two maps - one for http devices - httpDeviceMap, another one for mqtt - mqttDeviceMap. 
And remove this not so obvious halfSize idx manipulations